app/src/org/commcare/android/util/MediaUtil.java

```diff
@@ -1,82 +1,274 @@
+                                                   int calcHeight, int calcWidth,
+                                                   int boundingHeight, int boundingWidth) {
+
+        Pair<Integer, Integer> dimensImposedByContainer = getProportionalDimensForContainer(
+                originalHeight, originalWidth, boundingHeight, boundingWidth);
+        int newWidth = Math.min(dimensImposedByContainer.first, calcWidth);
+        int newHeight = Math.min(dimensImposedByContainer.second, calcHeight);
+
+        try {
+            BitmapFactory.Options o = new BitmapFactory.Options();
+            o.inScaled = false;
+            Bitmap originalBitmap = BitmapFactory.decodeFile(imageFilepath, o);
+            return Bitmap.createScaledBitmap(originalBitmap, newWidth, newHeight, false);
+        }
+        catch (OutOfMemoryError e) {
+            // OOM encountered trying to decode the bitmap at its current size, so we know we
+            // need to scale down by some factor greater than 1
+            int scaleFactor = originalHeight / newHeight;
+            if (scaleFactor < 2) {
+                scaleFactor = 2;
+            }
+            return performSafeScaleDown(imageFilepath, scaleFactor, 0);
+        }
+    }
+
+    /**
+     * @return A (width, height) pair representing the largest dimensions for which the aspect
+     * ratio given by originalHeight and originalWidth is maintained, without exceeding
+     * boundingHeight or boundingWidth
+     */
+    private static Pair<Integer, Integer> getProportionalDimensForContainer(int originalHeight,
+                                                                            int originalWidth,
+                                                                            int boundingHeight,
+                                                                            int boundingWidth) {
+        double heightScaleFactor = (double)boundingHeight / originalHeight;
+        double widthScaleFactor =  (double)boundingWidth / originalWidth;
+        double dominantScaleFactor = Math.min(heightScaleFactor, widthScaleFactor);
+
+        int widthImposedByContainer = (int)Math.round(originalWidth * dominantScaleFactor);
+        int heightImposedByContainer = (int)Math.round(originalHeight * dominantScaleFactor);
+        return new Pair<>(widthImposedByContainer, heightImposedByContainer);
+    }
+
+    /**
+     * @return A bitmap representation of the given image file, scaled up as close as possible to
+     * desiredWidth and desiredHeight, without exceeding either boundingHeight or boundingWidth
+     */
+    private static Bitmap attemptBoundedScaleUp(String imageFilepath, int desiredHeight, int desiredWidth,
+                                         int boundingHeight, int boundingWidth) {
+        if (boundingHeight < desiredHeight || boundingWidth < desiredWidth) {
+            float heightScale = ((float)boundingHeight) / desiredHeight;

```

Comment:
This pattern comes up a few times. Am wondering whether it would improve clarity to have it be a submethod